feature/preview/src/main/java/com/google/jetpackcamera/feature/preview/ui/CameraXViewfinder.kt

```diff
@@ -53,18 +59,25 @@ import kotlinx.coroutines.launch
 fun CameraXViewfinder(
     modifier: Modifier = Modifier,
     implementationMode: ImplementationMode = ImplementationMode.PERFORMANCE,
-    onSurfaceProviderReady: (Preview.SurfaceProvider) -> Unit = {}
+    lifecycleOwner: LifecycleOwner = LocalLifecycleOwner.current,
+    onSurfaceProviderReady: (Preview.SurfaceProvider) -> Unit = {},
+    onSurfaceProviderDisposed: () -> Unit = {}
 ) {
+    var surfaceProvider by remember { mutableStateOf<Preview.SurfaceProvider?>(null) }
+
+    LifecycleStartEffect(surfaceProvider, lifecycleOwner) {
+        surfaceProvider?.let { onSurfaceProviderReady(it) }
```

Comment:
Should we pull `surfaceProvider?.let` outside `LifecycleStartEffect`? It seems we'd be launching an extra side-effect for a `null` `SurfaceProvider` that will be cancelled immediately.

In fact, the more I think about this, maybe we don't even need a `State` variable for `surfaceProvider`. Calling `onSurfaceProviderDisposed()` when the `produceState` block completes or is cancelled seems to make the most sense to me. It wouldn't cause any unnecessary recompositions, and would actually pair `onSurfaceProviderReady` with `onSurfaceProviderDisposed` when the `implementationMode` changes.

Reply:
Thinking about this again, I still think it makes sense from an API perspective for `CameraXViewfinder` to add `onSurfaceProviderDisposed` to the existing `produceState`. However, I think that would only be for API completeness, and I don't think it would be used in JCA in its current state.

Ultimately, I don't think we should be pushing more state, including the lifecycle, into the `CameraXViewfinder` composable. It makes the surrounding code cleaner in the case of JCA, but doesn't necessarily represent what's actually going on: the `SurfaceProvider` lifecycle isn't actually linked to `lifecycleOwner` argument, though it behaves as if it is.